### PR TITLE
feat: fail install if any instance pending upgrade

### PR DESCRIPTION
**Pull request by fnaranjo-vmw**
[#185853346](https://www.pivotaltracker.com/story/show/185853346)

### Checklist:

* [ ] Have you added Release Notes in the docs respositories?
* [x] Have you followed the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)?

### CODE REVIEW

File: internal/upgrader/upgrader_test.go [hunk 1]
```diff
@@ -232,6 +232,98 @@ var _ = Describe("Upgrade", func() {
...
+			),
+			Entry("upgradeable instances whose creation failed",
+				nil, "no instances available to upgrade",
+				[]ccapi.Plan{{GUID: fakeInstance1.PlanGUID}},
+				[]ccapi.ServiceInstance{{UpgradeAvailable: true, LastOperation: ccapi.LastOperation{Type: "create", State: "failed"}}},
+			),
+			/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+			// WARNING:
+			// We may want to check whether the following scenarios are possible and if they do, we may want to prevent them from happening
+			/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (fnaranjo-vmw) [inline internal/upgrader/upgrader_test.go:312]:
```suggestion
```
Discussed these scenarios during standup and we are already investigating some possible corner cases.
Deleting this warning.

File: internal/upgrader/upgrader_test.go [hunk 2]
```diff
@@ -232,6 +232,98 @@ var _ = Describe("Upgrade", func() {
...
+			Entry("upgradeable instances whose creation failed and are in a possible invalid state?",
+				fmt.Errorf("check up-to-date failed: found 1 instances which are not up-to-date"),
+				`(\s|.)*`, // we don't care about the logs in this test
```
Comment (fnaranjo-vmw) [inline internal/upgrader/upgrader_test.go:315]:
```suggestion
			Entry("upgradeable instances after a failed destroy attempt",
				fmt.Errorf("check up-to-date failed: found 1 instances which are not up-to-date"),
				`(\s|.)*`, // we don't care about the logs in this test, only that it raises the error above
```

File: internal/upgrader/upgrader_test.go [hunk 3]
```diff
@@ -232,6 +232,98 @@ var _ = Describe("Upgrade", func() {
...
+				[]ccapi.Plan{{GUID: fakeInstance1.PlanGUID}},
+				[]ccapi.ServiceInstance{{UpgradeAvailable: true, LastOperation: ccapi.LastOperation{Type: "destroy", State: "failed"}}},
+			),
+			Entry("unlikely scenario in which the api returns instances which doesn't correspond to the queried plan",
+				fmt.Errorf("check up-to-date failed: found 1 instances which are not up-to-date"),
+				`(\s|.)*`, // we don't care about the logs in this test
+				[]ccapi.Plan{{GUID: "guid-for-unlikely-scenario"}},
+				[]ccapi.ServiceInstance{getFakeInstanceDetailed()},
+			),
```
Comment (fnaranjo-vmw) [inline internal/upgrader/upgrader_test.go:324]:
```suggestion
```

This test is unrealistic and doesn't provide real value.
This test is faking the result of two API calls, returning invalid data. So this test assumes API is behaving abnormally, which is a very unlikely scenario and out of our control.

However, we are already analysing a more realistic scenario:

When a Plan is deleted but it contains active instances Cloudfoundry disables the plan instead of deleting it
Disabled instances may contain a MaintenanceInfoVersion which is not representative of the current version of the Tile being installed (apparently, MaintenaceInfoVersion for disabled plans doesn't get updated)
Therefore, the mechanism for checking if instances need upgrading will determine that there are no upgrades available
After we finish analysing this corner-case we will make any required changes. For now, let's remove this useless test.

File: internal/config/config.go [hunk 4]
```diff
@@ -96,6 +102,8 @@ func Options() map[string]string {
 		parallelFlag:    parallelDescription,
 		httpLoggingFlag: httpLoggingDescription,
 		dryRunFlag:      dryRunDescription,
+
+		checkUpToDateFlag: checkUpToDateDescription,
```
[the remark this answers is not in the document]
Comment (zucchinidev) [reply internal/config/config.go:106]: This is only my preference. Feel free to ignore it.

File: internal/upgrader/upgrader.go [hunk 5]
```diff
@@ -27,7 +27,7 @@ type Logger interface {
 	FinalTotals()
 }
 
-func Upgrade(api CFClient, brokerName string, parallelUpgrades int, dryRun bool, log Logger) error {
+func Upgrade(api CFClient, brokerName string, parallelUpgrades int, dryRun bool, checkUpToDate bool, log Logger) error {
```
Comment (zucchinidev) [inline internal/upgrader/upgrader.go:30]:
```suggestion
func Upgrade(api CFClient, brokerName string, parallelUpgrades int, dryRun, checkUpToDate bool, log Logger) error {
```

File: internal/upgrader/upgrader.go [hunk 6]
```diff
@@ -47,6 +47,9 @@ func Upgrade(api CFClient, brokerName string, parallelUpgrades int, dryRun bool,
 	case len(upgradableInstances) == 0:
 		log.Printf("no instances available to upgrade")
 		return nil
+	case checkUpToDate:
+		log.InitialTotals(totalServiceInstances, len(upgradableInstances))
```
Comment (zucchinidev) [inline internal/upgrader/upgrader.go:51]: thoughts: the `log.InitialTotals` logs this sentence: `l.printf("upgradable instances: %d", upgradableInstances)`

I am not sure if it reflects the reality, I mean, are all of them upgradable?
In any case, I am happy with the modification and it seems reasonable

**Review by zucchinidev — APPROVED** [review]
